rfcs/InputUnion.md

```diff
@@ -18,8 +18,123 @@ To help bring this idea to reality, you can contribute through two channels:
+    animals {
+      __typename
+      name
+      age
+      ... on Cat { livesLeft }
+      ... on Dog { breed }
+      ... on Snake { venom }
+    }
+  }
+}
+```
+
+However, when we want to submit data, we can't use an `interface` or `union`, so we must model around that.
+
+One technique commonly used to is a **tagged union** pattern. This essentially boils down to a "wrapper" input that isolates each type into it's own field. The field name takes on the convention of representing the type.
+
+```graphql
+mutation {
+  logAnimalDropOff(
+    location: "Portland, OR"
+    animals: [
+      {cat: {name: "Buster", age: 3, livesLeft: 7}}
+    ]
+  )
+}
+```
+
+Unfortunately, this opens up a set of problems, since the Tagged union input type actually contains many fields, any of which could be submitted.
+
+```graphql
+input AnimalDropOffInput {
+  cat: CatInput
+  dog: DogInput
+  snake: SnakeInput
+}
+```
+
+This allows non-sensical mutations to pass GraphQL validation, for example representing an animal that is both a `Cat` and a `Dog`.
+
+```graphql
+mutation {
+  logAnimalDropOff(
+    location: "Portland, OR"
+    animals: [
+      {
+        cat: {name: "Buster", age: 3, livesLeft: 7},
+        dog: {name: "Ripple", age: 2, breed: WHIPPET}
+      }
+    ]
+  )
+}
+```
+
+In addition, relying on this layer of abstraction means that this domain must be modelled differently across input & output. This can put a larger burden on the developer interacting with the schema, both in terms of lines of code and complexity.
+
+```json
+// JSON structure returned from a query
+{
+  "animals": [
+    {"__typename": "Cat", "name": "Ruby", "age": 2, "livesLeft": 9}
+    {"__typename": "Snake", "name": "Monty", "age": 13, "venom": "POISON"}
+  ]
+}
+```
+
+```json
+// JSON structure submitted to a mutation
+{
+  "animals": [
+    {"cat": {"name": "Ruby", "age": 2, "livesLeft": 9}},
+    {"snake": {"name": "Monty", "age": 13, "venom": "POISON"}}
+  ]
+}
+```
+
+Another common approach is to provide a unique mutation for every type. A schema employing this technique might have `logCatDropOff`, `logDogDropOff` and `logSnakeDropOff` mutations. This removes the potential for modeling non-sensical situations, but it explodes the number of mutations in a schema, making the schema less accessible. If the type is nested inside other inputs, this approach simply isn't feasable.
+
+These workarounds only get worse at scale. Real world GraphQL schemas can have dozens if not hundreds of possible types for a single `Interface` or `Union`.
+
+The goal of the **Input Union** is to bring a polymorphic type to Inputs. This would enable us to model situations where an input may be of different types in a type-safe and elegant manner, like we can with outputs.
+
+```graphql
+mutation {
+  logAnimalDropOff(
+    location: "Portland, OR"
+    animals: [
+      {name: "Buster", age: 3, livesLeft: 7},
+      {name: "Ripple", age: 2, __typename: "DogInput"},
+    ]
+  )
+}
+```
```

Comment:
Using `__typename`, `__inputname` was one of the most controversial topics in the previous discussions so I don't think it worth having it in `Problem Sketch` section.
It can look like this instead:
--------------------
``` graphql
mutation {
  logAnimalDropOff(
    location: "Portland, OR"
    animals: [
      {name: "Buster", age: 3, livesLeft: 7},
      {name: "Ripple", age: 2},
    ]
  )
}
```
In this mutation, we encounter the main challenge of the **Input Union** - we need to determine the correct type of the data submitted.

A wide variety of solutions have been explored by the community, and they are outlined in detail in this document under [Possible Solutions](#Possible-Solutions).

-----------------------

Reply:
In my original editing I made this change, but it didn't feel sufficiently clear - I figured someone not familiar with the issue may not have detected the ambiguity. I then added comments to the end of each line `# This is meant to be CatInput` / `# This is meant to be DogInput` but it seemed messy and didn't really fix the issue. In the end I gave up and left it as the `__typename` solution.

However, I agree that if possible this should be addressed.

How about we add a comment `# Problem: the type of this Animal is ambiguous` to the second input?

```graphql
mutation {
  logAnimalDropOff(
    location: "Portland, OR"
    animals: [
      { name: "Buster", age: 3, livesLeft: 7 }
      { name: "Ripple", age: 2 } # Problem: the type of this Animal is ambiguous
    ]
  )
}
```

Reply:
@benjie Agree, comments works.
One note: type of both animals is ambiguous since `livesLeft` can become ambiguous if client extends `Dog` with `livesLeft` field or add new animal type with `livesLeft` in it. We can solve this by introducing some restrictions of having unique fields but it would part of solution and not problem definition.

So I would suggest:
```graphql
mutation {
  logAnimalDropOff(
    location: "Portland, OR",

    # Problem: we need to determinate type of the Animal for every item
    animals: [
      # This is meant to be CatInput
      { name: "Buster", age: 3, livesLeft: 7 }

      # This is meant to be DogInput
      { name: "Ripple", age: 2 }
    ]
  )
}
```

Reply:
^ Well written, @IvanGoncharov; I agree.

Reply:
@benjie @binaryseed I reviewed the entire document and everything looks great.
This is the last suggestion I have, so to clarify on final version:

--------------------
``` graphql
mutation {
  logAnimalDropOff(
    location: "Portland, OR",

    # Problem: we need to determinate type of the Animal for every item
    animals: [
      # This is meant to be CatInput
      { name: "Buster", age: 3, livesLeft: 7 },

      # This is meant to be DogInput
      { name: "Ripple", age: 2 },
    ]
  )
}
```
In this mutation, we encounter the main challenge of the **Input Union** - we need to determine the correct type of the data submitted.

A wide variety of solutions have been explored by the community, and they are outlined in detail in this document under [Possible Solutions](#Possible-Solutions).

-----------------------
I think mutation query now is self-explaining so we can skeep SDL part and don't start a conversation on `inputunion` vs `inputUnion` vs `@oneField` vs all other possible SDL syntaxes on this stage of RFC process.

Reply:
updated, great feedback!